### PR TITLE
Adding replace-command hook.

### PR DESCRIPTION
**Pull request by grasmash**

### CODE REVIEW

File: src/CommandProcessor.php [hunk 1]
```diff
@@ -137,6 +138,11 @@ public function validateRunAndAlter(
             return $validated;
         }
 
+        $replaceDispatcher = new ReplaceCommandHookDispatcher($this->hookManager(), $names);
+        if ($replaceDispatcher->hasReplaceCommandHook()) {
+            $commandCallback = $replaceDispatcher->getReplacementCommand($commandData);
```
Comment (greg-1-anderson) [inline src/CommandProcessor.php]: Nice!

File: src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * @param \Consolidation\AnnotatedCommand\CommandData $commandData
+     *
+     * @return callable
+     */
+    public function getReplacementCommand(CommandData $commandData)
+    {
+        $replaceCommandHooks = $this->getReplaceCommandHooks($commandData);
+
+        // We only take the first hook implementation of "replace-command" as the replacement. Commands shouldn't have
+        // more than one replacement.
```
Comment (greg-1-anderson) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: I wonder what the right UX should be if there are multiple replacements for the same command. Example: using an application that supports plugins, user installs two independently-developed plugins with overlapping functionality, both of which replace the same command. Perhaps we should simply log a warning when this happens, and print out the class::method names of every replacement for the given function. This would be as good a place as any to do that -- the desired behavior being that you don't get the warning unless you try to run a command with multiple replacement options.

**Comment by greg-1-anderson** [comment]
Looks great; could you add at least one test?

File: src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php (hunk 2 above)
Comment (grasmash) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: Does this class have access to the `$output`? I'd have to change the method signatures to pass output into `validateRunAndAlter()` and `getReplacementCommand()`  think. That ok?

Comment (greg-1-anderson) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: If you wanted to log in this method, you'd need to inject a logger into ReplaceCommandHookDispatcher. The CommandDispatcher has a reference to the logger that it could easily provide via constructor injection.

Alternately, you could log from the CommandDispatcher, and enhance the ReplaceCommandHookDispatcher to report whether there are zero, one or many hooks.

I don't have a preference which way this is implemented, or which hook fires when there is more than one replacement.

File: src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            $command_name = $commandData->annotationData()->get('command', 'unknown');
+            $output->writeln("<comment>Warning: multiple implementations of the \"replace-command\" hook exist for the \"$command_name\" command:</comment>");
+            foreach($replaceCommandHooks as $replaceCommandHook) {
+                $class = get_class($replaceCommandHook[0]);
+                $method = $replaceCommandHook[1];
+                $hook_name = "$class->$method";
+                $output->writeln("<comment>  - $hook_name</comment>");
+            }
+        }
+
```
Comment (grasmash) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: This results in output like:
```
Warning: multiple implementations of the "replace-command" hook exist for the "tests:behat" command:
  - Acquia\Blt\Robo\Hooks\ReplaceCommandHook->validateDocrootIsPresent
  - Acquia\Blt\Robo\Hooks\ReplaceCommandHook->validateDocrootIsPresent2
```

File: src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php (hunk 2 above)
Comment (grasmash) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: > The CommandDispatcher has a reference to the logger 

I assume you're referring the to CommandProcessor, but I do not see a reference to the logger in it. I'm implemented a solution that directly passed the `$output`.

Comment (greg-1-anderson) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: Yes, I did mean CommandProcessor, but I don't know what I was looking at that made me think there was a logger there, as none of the classes in this project log at the moment.

I don't want `$output` references here. To fix this, use [LoggerAwareTrait](https://github.com/php-fig/log/blob/master/Psr/Log/LoggerAwareTrait.php) (and LoggerAwareInterface, in the same project). You'd have to add it to CommandProcessor, and have CommandProcessor set the logger on the ReplaceCommandHookDispatcher when it was instantiated. You'd need to silently not log if no logger is provided (for backwards compatibility).

However, since there is no logging at all in this project ATM, it's also okay with me if you just back out the `$output` change, and leave logging out for now.

File: src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php (hunk 3 above)
Comment (greg-1-anderson) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: Looks great, but I don't want to use `$output` for logging, so please either back this out or replace with a logger.

File: src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php [hunk 4]
```diff
@@ -17,6 +18,15 @@ class ReplaceCommandHookDispatcher extends HookDispatcher
 {
 
     /**
+     * @var LoggerInterface|null
+     */
+    protected $logger = null;
+
+    public function setLogger(LoggerInterface $logger) {
```
Comment (greg-1-anderson) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: Should use LoggerAwareInterface / LoggerAwareTrait. The later is an optimization (avoid duplicate code), but the former is critical, as it will trigger dependency injection for classes in the DI container that implement it.

Comment (grasmash) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: I had assumed that ReplaceCommandHookDispatcher wasn't a container service and consequently would not receive the Logger via inflection.

Comment (grasmash) [inline src/Hooks/Dispatchers/ReplaceCommandHookDispatcher.php]: Ah I see. We can do it anyway.

File: src/CommandProcessor.php [hunk 5]
```diff
@@ -137,8 +141,19 @@ public function validateRunAndAlter(
             return $validated;
         }
 
-        // Run the command, alter the results, and then handle output and status
-        $result = $this->runCommandCallback($commandCallback, $commandData);
+        $replaceDispatcher = new ReplaceCommandHookDispatcher($this->hookManager(), $names);
+        if ($this->logger) {
```
Comment (greg-1-anderson) [inline src/CommandProcessor.php]: Minor style point: avoid unnecessary `if`s. It doesn't hurt to `setLogger()` with null, so there's no need to test here.

Comment (grasmash) [inline src/CommandProcessor.php]: That will not work since the `setLogger()` method type hints its first argument with `LoggerInterface`.

File: README.md [hunk 6]
```diff
@@ -195,6 +196,25 @@ class MyCommands implements CustomEventAwareInterface
...
+For instance, if you'd like to replace the `foo:bar` command, you could utilize the following code:
+
+```php
+<?php
+class MyReplaceCommandHook  {
+
+  /**
+   * @hook replace-command foo:bar
+   */
+  public function myFooBarReplacement($args) {
```
Comment (grasmash) [inline README.md]: Actually, I had to modify things so that they are all passed as a single array of arguments.

File: src/CommandProcessor.php (hunk 5 above)
Comment (grasmash) [inline src/CommandProcessor.php]: Er, well that's not true. I guess we can just have it match? I liked using a single array, but we can write less code the other way. Let me know which you'd prefer.

File: README.md (hunk 6 above)
Comment (grasmash) [inline README.md]: Er, well that's not true. I guess we can just have it match? I liked using a single array, but we can write less code the other way. Let me know which you'd prefer.

File: src/CommandProcessor.php [hunk 7]
```diff
@@ -137,8 +141,19 @@ public function validateRunAndAlter(
...
+            $replaceDispatcher->setLogger($this->logger);
+        }
+        if ($replaceDispatcher->hasReplaceCommandHook()) {
```
Comment (greg-1-anderson) [inline src/CommandProcessor.php]: Why did you add this extra code? It was much nicer the way it was before. A command replacement hook can only every replace one command, so the function arguments will always be well-known. Just use the same signature as the calling function, per your original implementation. The code is cleaner, and the UX is better.

File: tests/src/ExampleCommandFile.php [hunk 8]
```diff
@@ -256,6 +256,22 @@ public function alterMeToo()
...
+     */
+    public function testReplaceCommand($value)
+    {
+        $this->output->writeln($value);
+    }
+
+    /**
+     * @hook replace-command test:replace-command
+     */
+    public function hookTestReplaceCommandHook(array $args)
```
Comment (greg-1-anderson) [inline tests/src/ExampleCommandFile.php]: So this should just be `$value`

File: README.md (hunk 6 above)
Comment (grasmash) [inline README.md]: Reverted

File: src/CommandProcessor.php (hunk 7 above)
Comment (grasmash) [inline src/CommandProcessor.php]: Because my tests were failing and I resolved it the wrong way. Reverted.

**Comment by greg-1-anderson** [comment]
Great -- thanks for sticking with it & adding tests and logger support.
